vscode/src/completions/utils/lexical-analysis.ts

```diff
@@ -0,0 +1,161 @@
+ * List of all supported languages that we have grammars and
+ * lexems for. Note that enum values are copied from VSCode API,
+ * if we want to make it work with different editors we should
+ * enhance language detection.
+ *
+ * TODO: Decouple language detect to make it editor agnostic
+ */
+export enum SupportedLanguage {
+    JavaScript = 'javascript',
+    TypeScript = 'typescript',
+    Java = 'java',
+    Go = 'go',
+    // The problem with these two is that we have to use typescript
+    // wasm module grammar, the problem that exports in wasm works differently
+    // we probably have to compile custom wasm module for tsx grammar.
+    // JSX = 'javascriptreact',
+    // TSX = 'typescriptreact',
+}
+
+// TODO: Add grammar type autogenerate script
+// see https://github.com/asgerf/dts-tree-sitter
+type GrammarPath = string
+
+const SUPPORTED_LANGUAGES: Record<SupportedLanguage, GrammarPath> = {
+    [SupportedLanguage.JavaScript]: 'tree-sitter-javascript.wasm',
+    [SupportedLanguage.Java]: 'tree-sitter-java.wasm',
+    [SupportedLanguage.Go]: 'tree-sitter-go.wasm',
+    [SupportedLanguage.TypeScript]: 'tree-sitter-typescript.wasm',
+    // Since TypeScript is a subset over javascript grammar we
+    // use typescript grammar here in order to support jsx parsing
+    // [SupportedLanguage.JSX]: TypeScript.tsx,
+    // [SupportedLanguage.TypeScript]: TypeScript.typescript,
+    // [SupportedLanguage.TSX]: TypeScript.tsx,
+}
+
+export enum GenericLexem {
+    IfStatement,
+    ElseClause,
+    StatementBlock,
+    CallExpression,
+}
+
+enum JavaScriptLexemType {
+    IfStatement = 'if_statement',
+    ElseClause = 'else_clause',
+    StatementBlock = 'statement_block',
+    CallExpression = ' call_expression',
+}
+
+enum JavaLexemType {
+    IfStatement = 'if_statement',
+    ElseClause = 'else',
+    StatementBlock = 'block',
+    MethodInvocation = 'method_invocation',
+}
+
+enum GoLexemType {
+    IfStatement = 'if_statement',
+    ElseClause = 'else',
+    StatementBlock = 'block',
+    CallExpression = 'call_expression',
+}
+
+type LEXEME_DICTIONARY = Record<GenericLexem, string>
+
+export const LANGUAGE_TO_LEXEM: Partial<Record<SupportedLanguage, LEXEME_DICTIONARY>> = {
+    [SupportedLanguage.JavaScript]: {
+        [GenericLexem.IfStatement]: JavaScriptLexemType.IfStatement,
+        [GenericLexem.ElseClause]: JavaScriptLexemType.ElseClause,
+        [GenericLexem.StatementBlock]: JavaScriptLexemType.StatementBlock,
+        [GenericLexem.CallExpression]: JavaScriptLexemType.CallExpression,
+    },
+    // We reuse JavaScript lexemes for typescript since TS grammar extends
+    // JavaScript grammar
+    [SupportedLanguage.TypeScript]: {
+        [GenericLexem.IfStatement]: JavaScriptLexemType.IfStatement,
+        [GenericLexem.ElseClause]: JavaScriptLexemType.ElseClause,
+        [GenericLexem.StatementBlock]: JavaScriptLexemType.StatementBlock,
+        [GenericLexem.CallExpression]: JavaScriptLexemType.CallExpression,
+    },
```

Comment:
There's a lot of mapping going on. What if we just have one interface `GenericLexem` and then return an object here to match that like so?

```suggestion
    'javascript':
    // We reuse JavaScript lexemes for typescript since TS grammar extends
    // JavaScript grammar
    'typescript: {
        ifStatement: 'if_statement',
        elseClause: 'else_clause',
        statementBlock: 'statement_block',
        callExpression: 'call_expression',
    },
```

So we can also remove the interim `JavaScriptLexemType` enum?